File: typescript-playwright/src/clients/CommentClient.ts
import type { CommentApi } from "../api/CommentApi";
import type {
  CommentDto,
  CaptchaData,
  CaptchaResponse,
  CommentsPageResponse,
  CommentPayload,
} from "../utils/types";
import { statusIn } from "../utils/statusIn";
import { HTTP } from "../utils/constants";
import { step } from "../utils/decorators";
import { generate } from "../utils/generators";

export class CommentClient {
  private readonly api: CommentApi;

  constructor(api: CommentApi) {
    this.api = api;
  }

  @step()
  async getCaptcha(): Promise<CaptchaResponse> {
    const response = await this.api.getCaptcha();
    if (!statusIn(HTTP.OK)(response.status())) {
      throw new Error(
        `[CommentClient] getCaptcha failed: ${response.status()} ${await response.text()}`,
      );
    }
    return response.json() as Promise<CaptchaResponse>;
  }

  @step()
  async listByPhoto(
    photoId: number,
    page = 0,
    size = 5,
  ): Promise<CommentsPageResponse> {
    const response = await this.api.getComments(photoId, page, size);
    if (!statusIn(HTTP.OK)(response.status())) {
      throw new Error(
        `[CommentClient] listByPhoto failed: ${response.status()} ${await response.text()}`,
      );
    }
    return response.json() as Promise<CommentsPageResponse>;
  }

  @step()
  async create(
    payload: CommentPayload,
    captcha: CaptchaData,
  ): Promise<CommentDto> {
    const response = await this.api.create(payload, captcha);
    if (!statusIn(HTTP.OK, HTTP.CREATED)(response.status())) {
      throw new Error(
        `[CommentClient] create failed: ${response.status()} ${await response.text()}`,
      );
    }
    return response.json() as Promise<CommentDto>;
  }

  @step()
  async createInIsolation(
    payload: CommentPayload,
    captcha: CaptchaData,
  ): Promise<CommentDto> {
    const response = await this.api.create(payload, captcha, {
      "X-Forwarded-For": generate.ip(),
    });
    if (!statusIn(HTTP.OK, HTTP.CREATED)(response.status())) {
      throw new Error(
        `[CommentClient] createInIsolation failed: ${response.status()} ${await response.text()}`,
      );
    }
    return response.json() as Promise<CommentDto>;
  }
}

File: python-pytest/src/clients/comment_client.py
from __future__ import annotations

from typing import cast

from src.api.comment_api import CommentApi
from src.utils.constants import HTTP
from src.utils.decorators import step
from src.utils.types import (
    CaptchaData,
    CaptchaResponse,
    CommentDto,
    CommentPayload,
    CommentsPageResponse,
)


class CommentClient:
    def __init__(self, api: CommentApi) -> None:
        self._api = api

    @step()
    def get_captcha(self) -> CaptchaResponse:
        response = self._api.get_captcha()
        if response.status_code not in (HTTP.OK,):
            raise RuntimeError(
                f"[CommentClient] getCaptcha failed: {response.status_code} {response.text}"
            )
        return cast(CaptchaResponse, response.json())

    @step()
    def list_by_photo(self, photo_id: int, page: int = 0, size: int = 5) -> CommentsPageResponse:
        response = self._api.get_comments(photo_id, page, size)
        if response.status_code not in (HTTP.OK,):
            raise RuntimeError(
                f"[CommentClient] listByPhoto failed: {response.status_code} {response.text}"
            )
        return cast(CommentsPageResponse, response.json())

    @step()
    def create(self, payload: CommentPayload, captcha: CaptchaData) -> CommentDto:
        response = self._api.create(payload, captcha)
        if response.status_code not in (HTTP.OK, HTTP.CREATED):
            raise RuntimeError(
                f"[CommentClient] create failed: {response.status_code} {response.text}"
            )
        return cast(CommentDto, response.json())

    @step()
    def create_in_isolation(self, payload: CommentPayload, captcha: CaptchaData) -> CommentDto:
        from src.utils.generators import Generate

        response = self._api.create(payload, captcha, {"X-Forwarded-For": Generate.ip()})
        if response.status_code not in (HTTP.OK, HTTP.CREATED):
            raise RuntimeError(
                f"[CommentClient] createInIsolation failed: {response.status_code} {response.text}"
            )
        return cast(CommentDto, response.json())
